whois-api/src/test/java/net/ripe/db/whois/api/rest/WhoisRestServiceDoSTestIntegration.java

```diff
@@ -0,0 +1,289 @@
+        allOf.join();
+
+        // Collect status
+        final List<Integer> statuses = futures.stream()
+                .map(future -> {
+                    try {
+                        return future.get().getStatus();
+                    } catch (Exception e) {
+                        return null;
+                    }
+                })
+                .toList();
+        final List<Integer> manyRequestStatuses = statuses.stream().filter(status -> HttpStatus.TOO_MANY_REQUESTS_429 == status).toList();
+        assertThat(manyRequestStatuses.size(), is(1));
+
+        TimeUnit.SECONDS.sleep(1); // Free the IP after one second
+
+        //After a second, the user can perform more requests
+        final Response unLockedResponse = RestTest.target(getPort(), "whois/test/mntner/OWNER-MNT?clientIp=10.20.30.40")
+                .request()
+                .get();
+
+        assertThat(HttpStatus.OK_200, is(unLockedResponse.getStatus()));
+    }
+
+    @Test
+    public void multiple_lookUps_per_second_but_white_list_IP_then_200() {
+        final Response response = RestTest.target(getPort(), "whois/test/mntner/OWNER-MNT")
+                .request()
+                .get();
+        assertThat(HttpStatus.OK_200, is(response.getStatus()));
+
+        final Invocation.Builder lookupRequest = RestTest.target(getPort(), "whois/test/mntner/OWNER-MNT").request();
+        final List<Integer> statuses = Lists.newArrayList();
+        for (int i = 0; i < 50; i++) {
+            statuses.add(lookupRequest.get().getStatus());
+        }
+
+        assertThat(HttpStatus.OK_200, is(statuses.getLast()));
+    }
+
+    // Updates
+    @Test
+    public void multiple_updates_per_second_then_429_too_many_requests() throws InterruptedException {
+        final RpslObject person = RpslObject.parse("" +
+                "person:    Pauleth Palthen\n" +
+                "address:   Singel 258\n" +
+                "phone:     +31-1234567890\n" +
+                "e-mail:    noreply@ripe.net\n" +
+                "mnt-by:    OWNER-MNT\n" +
+                "nic-hdl:   PP1-TEST\n" +
+                "remarks:   remark\n" +
+                "source:    TEST\n");
+
+        final Response response = RestTest.target(getPort(), "whois/test/person?clientIp=10.20.30.40&password=test")
+                .request()
+                .post(Entity.entity(map(person), MediaType.APPLICATION_XML));
+        assertThat(HttpStatus.OK_200, is(response.getStatus()));
+
+        final Invocation.Builder updateRequest = RestTest.target(getPort(), "whois/test/person/PP1-TEST?clientIp=10.20.30.40&password=test").request();
+        final List<Integer> statuses = Lists.newArrayList();
+        for (int i = 0; i < 10; i++) {
+            statuses.add(updateRequest.put(Entity.entity(map(person), MediaType.APPLICATION_XML)).getStatus());
+        }
+
+        assertThat(HttpStatus.OK_200, is(statuses.get(8)));
```

Comment:
There are a lot of magic numbers in the tests. Can you make it more generic and only check that *some* requests have failed due to the rate limit?

Reply:
These test are doing more than 10 updates
testing that the update 9 is 200 but the update 10 is 429. I can rename that 8 and put "penultimateUpdate"